### PR TITLE
fix(compiler): throw an error for invalid provider

### LINKED ISSUE #8870: Injector allows binding undefined, `provide` function does not fail when incorrectly used
- **I'm submitting a ...**
- [x] bug report

**Current behavior**

It's very easy to get incorrect behaviour out of the injector, e.g. by accidentally forgetting to pass an object literal to `provide`. For example, the following will silently fail and bind `undefined` to `MyFoo`.

``` ts
provide(MyFoo, MyFooImpl)  // should be: {useClass: MyFooImpl}
```

**Expected/desired behavior**

This should fail on two levels:
- `provide` should check its arguments, and if none of the arguments are defined, given an error.
- it should be impossible to bind `undefined`, and probably also impossible to bind `null`.

_`undefined`_ is a common error (some import or property is missing, something went wrong)

Allowing to bind `null` means the code that gets instantiated by the injector gets a ctor parameter (field, whatever) that does not satisfy the contract the injected code expects; doing anything with it will fail. This means very hard to debug error cases, where once the app is running, much later something somewhere fails. Guice handles this by requiring bindings to be annotated as `@Nullable` if they should take null, but I think we might as well just disallow it; passing `null` values long distance through a large software system is prone to create very hard to debug error cases.

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Closes #8870

### CODE REVIEW

File: modules/@angular/compiler/src/metadata_resolver.ts [hunk 1]
```diff
@@ -812,6 +813,19 @@ export class CompileMetadataResolver {
     return compileProviders;
   }
 
+  private _validate(provider: any): void {
+    if (provider.hasOwnProperty('useClass') && provider.useClass == null) {
+      this._reportError(new SyntaxError(
+          `Invalid provider for ${stringifyType(provider.provide)}. useClass cannot be null or undefined, got: ${provider.useClass}.
```
Comment (vicb) [inline modules/@angular/compiler/src/metadata_resolver.ts]: useClass cannot ${provider.useClass}.

File: modules/@angular/compiler/src/metadata_resolver.ts [hunk 2]
```diff
@@ -812,6 +813,19 @@ export class CompileMetadataResolver {
...
+           Usually it happens when:
+           1. There's a circular dependency (might be caused by using index.ts files).
+           2. Class was declared after it was used: 
+
+           const provider = {provide: SomeClass, useClass: SomeClass};
```
Comment (vicb) [inline modules/@angular/compiler/src/metadata_resolver.ts]: add `use forwardRef`

**Review by vicb — COMMENTED** [review]
Thanks.
Could you please fix the few comments

**Comment by DzmitryShylovich** [comment]
@vicb 
> Could you please fix the few comments

fixed

File: modules/@angular/compiler/src/metadata_resolver.ts [hunk 3]
```diff
@@ -15,7 +15,7 @@ import {DirectiveNormalizer} from './directive_normalizer';
 import {DirectiveResolver} from './directive_resolver';
 import {ListWrapper, StringMapWrapper} from './facade/collection';
```
Comment (vicb) [inline modules/@angular/compiler/src/metadata_resolver.ts]: seems like those are unused

**Comment by DzmitryShylovich** [comment]
@vicb 
> seems like those are unused

yeah, removed it

**Pull request merged**